File: 1.2/BinaryNotes.NET/org/bn/coders/Decoder.cs
/*
* Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
* 
* Licensed under the LGPL, Version 2 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* 
*      http://www.gnu.org/copyleft/lgpl.html
* 
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* 
* With any your questions welcome to my e-mail 
* or blog at http://abdulla-a.blogspot.com.
*/
using System;
using System.Reflection;
using System.IO;
using org.bn.attributes;

namespace org.bn.coders
{	
	public abstract class Decoder : IDecoder
	{		
        public virtual T decode<T>(Stream stream) {
            Type objectClass = typeof(T);
            ElementInfo elemInfo = 
                new ElementInfo(
                    objectClass,
                    CoderUtils.getAttribute<ASN1Element>(objectClass)
                );
            return (T)decodeClassType(decodeTag(stream),objectClass,elemInfo, stream).Value;
        }
		
		protected virtual DecodedObject<object> decodeClassType(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream)
		{
            if (elementInfo.isAttributePresent<ASN1SequenceOf>())
            {
                return decodeSequenceOf(decodedTag, objectClass, elementInfo, stream);
            }
            else
            if (elementInfo.isAttributePresent<ASN1Sequence>())
			{
				return decodeSequence(decodedTag, objectClass, elementInfo, stream);
			}
            else
            if (elementInfo.isAttributePresent<ASN1Choice>())
			{
				return decodeChoice(decodedTag, objectClass, elementInfo, stream);
			}
			else
            if (elementInfo.isAttributePresent<ASN1BoxedType>())
			{
				return decodeBoxedType(decodedTag, objectClass, elementInfo, stream);
			}
			else
            if (elementInfo.isAttributePresent<ASN1Enum>())
			{
				return decodeEnum(decodedTag, objectClass, elementInfo, stream);
			}
			else
            if (elementInfo.isAttributePresent<ASN1Boolean>())
			{
				return decodeBoolean(decodedTag, objectClass, elementInfo, stream);
			}
			else
            if (elementInfo.isAttributePresent<ASN1Any>())
			{
				return decodeAny(decodedTag, objectClass, elementInfo, stream);
			}
			else
            if (elementInfo.isAttributePresent<ASN1Integer>())
			{
				return decodeInteger(decodedTag, objectClass, elementInfo, stream);
			}
            else
            if (elementInfo.isAttributePresent<ASN1Real>())
            {
                return decodeReal(decodedTag, objectClass, elementInfo, stream);
            }
			else
            if (elementInfo.isAttributePresent<ASN1OctetString>())
			{
				return decodeOctetString(decodedTag, objectClass, elementInfo, stream);
			}
			else
            if (elementInfo.isAttributePresent<ASN1BitString>())
            {
                return decodeBitString(decodedTag, objectClass, elementInfo, stream);
            }
            else
            if (elementInfo.isAttributePresent<ASN1String>())
			{
				return decodeString(decodedTag, objectClass, elementInfo, stream);
			}
			else
            if (elementInfo.isAttributePresent<ASN1Null>())
			{
				return decodeNull(decodedTag, objectClass, elementInfo, stream);
			}
			else
            if (elementInfo.isAttributePresent<ASN1Element>())
			{
				return decodeElement(decodedTag, objectClass, elementInfo, stream);
			}
			else
				return decodeJavaElement(decodedTag, objectClass, elementInfo, stream);
		}
		
		protected virtual DecodedObject<object> decodeJavaElement(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream)
		{
			if (elementInfo.AnnotatedClass.Equals(typeof(string)))
			{
				return decodeString(decodedTag, objectClass, elementInfo, stream);
			}
			else if (elementInfo.AnnotatedClass.Equals(typeof(int)))
			{
				return decodeInteger(decodedTag, objectClass, elementInfo, stream);
			}
            else if (elementInfo.AnnotatedClass.Equals(typeof(long)))
            {
                return decodeInteger(decodedTag, objectClass, elementInfo, stream);
            }
            else if (elementInfo.AnnotatedClass.Equals(typeof(double)))
            {
                return decodeReal(decodedTag, objectClass, elementInfo, stream);
            }
			else if (elementInfo.AnnotatedClass.Equals(typeof(bool)))
			{
				return decodeBoolean(decodedTag, objectClass, elementInfo, stream);
			}
			/*if(objectClass.isArray()) {
			return decodeOctetString(elementInfo, stream);
			}
			else*/
			else
				return null;
		}
					
		protected virtual void  invokeSetterMethodForField(PropertyInfo field, object obj, object param)
		{
            field.SetValue(obj, param, null);
		}

        protected virtual MethodInfo findSelectMethodForField(PropertyInfo field, object obj)
        {
            string methodName = "select" + field.Name.ToUpper().Substring(0, (1) - (0)) + field.Name.Substring(1);
            return obj.GetType().GetMethod(methodName);
        }

        protected virtual void invokeSelectMethodForField(PropertyInfo field, object obj, object param)
        {
            MethodInfo method = findSelectMethodForField(field, obj);
            method.Invoke(obj, new object[] { param });
        }

        protected bool isOptionalField(PropertyInfo field)
        {
            if (CoderUtils.isAttributePresent<ASN1Element>(field))
            {
                ASN1Element info = CoderUtils.getAttribute<ASN1Element>(field);
                if (info.IsOptional || info.HasDefaultValue )
                    return true;
            }
            return false;
        }


        protected virtual void checkForOptionalField(PropertyInfo field)
        {
            if(!isOptionalField(field))
                throw new System.ArgumentException("The mandatory field '" + field.Name + "' does not have a value!");
        }

        protected virtual void initDefaultValues(object obj) {
            try {
                string methodName = "initWithDefaults";
                MethodInfo method = obj.GetType().GetMethod(methodName);
                method.Invoke(obj,null);
            }
            catch(Exception ){};
        }

		
		protected virtual DecodedObject<object> decodeSequence(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream)
		{
			object sequence = Activator.CreateInstance(objectClass);
            initDefaultValues(sequence);
			DecodedObject<object> fieldTag = decodeTag(stream);
			int sizeOfSequence = 0;
			if (fieldTag != null)
				sizeOfSequence += fieldTag.Size;
			PropertyInfo[] fields = 
                objectClass.GetProperties();
			for (int i = 0; i < fields.Length; i++)
			{
				PropertyInfo field = fields[i];
				DecodedObject<object> obj = decodeSequenceField(
                    fieldTag, sequence, field, stream, elementInfo,true
                );
				if (obj != null)
				{
					sizeOfSequence += obj.Size;
					
					if (i != fields.Length - 1 && CoderUtils.isAttributePresent<ASN1Any>(fields[i + 1]))
					{
					}
					else
					{
						fieldTag = decodeTag(stream);
						if (fieldTag != null)
							sizeOfSequence += fieldTag.Size;
						else
						{
							break;
						}
					}
				}
				;
			}
			return new DecodedObject<object>(sequence, sizeOfSequence);
		}
		
		protected virtual DecodedObject<object> decodeSequenceField(DecodedObject<object> fieldTag, object sequenceObj, PropertyInfo field, System.IO.Stream stream, ElementInfo elementInfo, bool checkOptional)
		{
            ElementInfo info = new ElementInfo(
                field,
                CoderUtils.getAttribute<ASN1Element>(field)
            );
            
            //info.setGenericInfo(field.getGenericType());
			if (CoderUtils.isAttributePresent<ASN1Null>(field))
			{
				return decodeNull(fieldTag, field.PropertyType, info, stream);
			}
			else
			{
                DecodedObject<object> val = decodeClassType(fieldTag, field.PropertyType, info, stream);
				if (val != null)
				{
					invokeSetterMethodForField(field, sequenceObj, val.Value);
				}
				else
				{
                    if(checkOptional)
					    checkForOptionalField(field);
				}
				return val;
			}
		}
		
		protected virtual DecodedObject<object> decodeChoice(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream)
		{
			object choice = System.Activator.CreateInstance(objectClass);
			DecodedObject<object> val = null;
			foreach(PropertyInfo field in objectClass.GetProperties())
			{
                ElementInfo info = new ElementInfo(
                    field,
                    CoderUtils.getAttribute<ASN1Element>(field)
                );
                //info.setGenericInfo(field.getGenericType());
                val = decodeClassType(decodedTag, field.PropertyType, info, stream);
				if (val != null)
				{
					invokeSelectMethodForField(field, choice, val.Value);
					break;
				}
				;
			}
			if (val == null && elementInfo.Element != null && !elementInfo.Element.IsOptional)
			{
				throw new System.ArgumentException("The choice '" + objectClass.ToString() + "' does not have a selected item!");
			}
			else
				return new DecodedObject<object>(choice, val != null?val.Size:0);
		}
		
		protected virtual DecodedObject<object> decodeEnum(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream)
		{
			object result = System.Activator.CreateInstance(objectClass);
            Type enumClass = null;
            foreach (MemberInfo member in objectClass.GetMembers())
            {
                if (member is System.Type)
                {
                    Type cls = (Type)member;
                    if (cls.IsEnum)
                    {
                        enumClass = cls;
                        break;
                    }
                }
            };

			System.Reflection.PropertyInfo field = objectClass.GetProperty("Value");


			DecodedObject<object> itemValue = decodeEnumItem(decodedTag, field.PropertyType, enumClass, elementInfo, stream);
			
            System.Reflection.FieldInfo param = null;
			if (itemValue != null)
			{
		        foreach(FieldInfo enumItem in enumClass.GetFields())
		        {
                    if (CoderUtils.isAttributePresent<ASN1EnumItem>(enumItem))
			        {
					    ASN1EnumItem meta =
                            CoderUtils.getAttribute<ASN1EnumItem>(enumItem);
					    if (meta.Tag.Equals(itemValue.Value))
					    {
						    param = enumItem;
						    break;
					    }
			        }
		        }
				invokeSetterMethodForField(field, result, param.GetValue(null));
			}
			return new DecodedObject<object>(result, itemValue.Size);
		}
		
		protected abstract DecodedObject<object> decodeEnumItem(DecodedObject<object> decodedTag, System.Type objectClass, System.Type enumClass, ElementInfo elementInfo, System.IO.Stream stream);
		
		
		protected virtual DecodedObject<object> decodeElement(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream)
		{
			elementInfo.AnnotatedClass = objectClass;
			return decodeClassType(decodedTag, objectClass, elementInfo, stream);
		}
		
		protected virtual DecodedObject<object> decodeBoxedType(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream)
		{
			object resultObj = System.Activator.CreateInstance(objectClass);
			
			DecodedObject<object> result = new DecodedObject<object>(resultObj);
			
			PropertyInfo field = objectClass.GetProperty("Value");
			elementInfo.AnnotatedClass = field;
			//elementInfo.setGenericInfo(field.getGenericType());
			
			DecodedObject<object> val = null;
            if (CoderUtils.isAttributePresent<ASN1Null>(field))
            {
				val = decodeNull(decodedTag, field.PropertyType, elementInfo, stream);
			}
			else
			{
                val = decodeClassType(decodedTag, field.PropertyType, elementInfo, stream);
				if (val != null)
				{
					result.Size = val.Size;
					invokeSetterMethodForField(field, resultObj, val.Value);
				}
				else
					result = null;
			}
			return result;
		}

        protected abstract DecodedObject<object> decodeBoolean(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream);

        protected abstract DecodedObject<object> decodeAny(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream);

        protected abstract DecodedObject<object> decodeNull(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream);

        protected abstract DecodedObject<object> decodeInteger(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream);

        protected abstract DecodedObject<object> decodeReal(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream);

        protected abstract DecodedObject<object> decodeOctetString(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream);

        protected abstract DecodedObject<object> decodeBitString(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream);

        protected abstract DecodedObject<object> decodeString(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream);

        protected abstract DecodedObject<object> decodeSequenceOf(DecodedObject<object> decodedTag, System.Type objectClass, ElementInfo elementInfo, System.IO.Stream stream);

        protected abstract DecodedObject<object> decodeTag(System.IO.Stream stream);


	}
}

File: 1.3-experimental-perf/JavaLibrary/src/org/bn/coders/Decoder.java
/*
 * Copyright 2006 Abdulla G. Abdurakhmanov (abdulla.abdurakhmanov@gmail.com).
 * 
 * Licensed under the LGPL, Version 2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.gnu.org/copyleft/lgpl.html
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 * With any your questions welcome to my e-mail 
 * or blog at http://abdulla-a.blogspot.com.
 */
package org.bn.coders;

import java.io.InputStream;

import java.lang.reflect.AnnotatedElement;

import java.lang.reflect.Constructor;

import org.bn.IDecoder;
import java.lang.reflect.Field;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Type;

import org.bn.annotations.*;


public abstract class Decoder implements IDecoder {    
    public <T> T decode(InputStream stream, Class<T> objectClass) throws Exception {

        ElementInfo elemInfo = 
            new ElementInfo(
                objectClass, 
                objectClass.getAnnotation(ASN1Element.class)
            );
        return (T)decodeClassType(decodeTag(stream),objectClass,elemInfo, stream).getValue();
    }
    
    protected DecodedObject decodeClassType(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception {
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1SequenceOf.class) ) {
            return decodeSequenceOf(decodedTag, objectClass,elementInfo, stream);
        }        
        else    
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1Sequence.class) ) {
            return decodeSequence(decodedTag,objectClass,elementInfo, stream);
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1Choice.class) ) {
            return decodeChoice(decodedTag,objectClass,elementInfo, stream);
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1BoxedType.class) ) {
            return decodeBoxedType(decodedTag,objectClass,elementInfo, stream);
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1Enum.class) ) {
            return decodeEnum(decodedTag, objectClass,elementInfo, stream);
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1Boolean.class) ) {            
            return decodeBoolean(decodedTag, objectClass,elementInfo, stream);
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1Any.class) ) {
            return decodeAny(decodedTag, objectClass,elementInfo, stream);
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1Integer.class) ) {
            return decodeInteger(decodedTag, objectClass,elementInfo, stream);
        }        
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1Real.class) ) {
            return decodeReal(decodedTag, objectClass,elementInfo, stream);
        }        
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1OctetString.class) ) {
            return decodeOctetString(decodedTag, objectClass,elementInfo, stream);
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1BitString.class) ) {
            return decodeBitString(decodedTag, objectClass,elementInfo, stream);
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1String.class) ) {
            return decodeString(decodedTag, objectClass,elementInfo, stream);
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1Null.class) ) {
            return decodeNull(decodedTag, objectClass,elementInfo, stream);        
        }
        else
        if( elementInfo.getAnnotatedClass().isAnnotationPresent(ASN1Element.class) ) {
            return decodeElement(decodedTag, objectClass,elementInfo, stream);
        }
        else
            return decodeJavaElement(decodedTag, objectClass,elementInfo, stream);
    }
    
    protected DecodedObject decodeJavaElement(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream ) throws Exception {
        if(elementInfo.getAnnotatedClass().equals(String.class)) {
            return decodeString(decodedTag, objectClass,elementInfo, stream);
        }
        else
        if(elementInfo.getAnnotatedClass().equals(Integer.class)) {
            return decodeInteger(decodedTag, objectClass,elementInfo, stream);
        }
        else
        if(elementInfo.getAnnotatedClass().equals(Long.class)) {
            return decodeInteger(decodedTag, objectClass,elementInfo, stream);
        }
        else
        if(elementInfo.getAnnotatedClass().equals(Double.class)) {
            return decodeReal(decodedTag, objectClass,elementInfo, stream);
        }
        else        
        if(elementInfo.getAnnotatedClass().equals(Boolean.class)) {
            return decodeBoolean(decodedTag, objectClass,elementInfo, stream);
        }        
        else
        /*if(objectClass.isArray()) {
            return decodeOctetString(elementInfo, stream);
        }
        else*/
            return null;
    }
    
    protected Method findMethodForField(String methodName, Object object, Object param) throws NoSuchMethodException {
        try {
            return object.getClass().getMethod(methodName, new Class[] {param.getClass()});
        }
        catch(NoSuchMethodException ex) {
            Method[] methods = object.getClass().getMethods();
            for(Method method : methods) {
                if(method.getName().equalsIgnoreCase(methodName)) {
                    return method;
                }
            }
            throw ex;
        }
    }    
    
    protected Method findSetterMethodForField(Field field, Object object, Object param) throws NoSuchMethodException {
        String methodName = "set"+field.getName().toUpperCase().substring(0,1)+field.getName().substring(1);
        return findMethodForField(methodName, object, param);
    }
    
    protected void invokeSetterMethodForField(Field field, Object object, Object param) throws Exception {
        Method method = findSetterMethodForField(field,object, param);
        method.invoke(object, param);
    }

    protected Method findSelectedMethodForField(Field field, Object object,Object param) throws NoSuchMethodException {
        String methodName = "select"+field.getName().toUpperCase().substring(0,1)+field.getName().substring(1);
        return findMethodForField(methodName, object, param);
    }

    protected void invokeSelectMethodForField(Field field, Object object, Object param) throws NoSuchMethodException, 
                                                                      IllegalAccessException, 
                                                                      InvocationTargetException {
        Method method = findSelectedMethodForField(field,object, param);
        method.invoke(object, param);
    }
    
    protected boolean isOptionalField(Field field) {
        if( field.isAnnotationPresent(ASN1Element.class) ) {
            ASN1Element info = field.getAnnotation(ASN1Element.class);
            if(info.isOptional() || info.hasDefaultValue())
                return true;
        }        
        return false;
    }
    
    
    protected void checkForOptionalField(Field field) throws Exception {
        if( isOptionalField(field) )
                return;
        throw new  IllegalArgumentException ("The mandatory field '" + field.getName() + "' does not have a value!");
    }
    
    protected void initDefaultValues(Object object) throws NoSuchMethodException, 
                                                           IllegalAccessException, 
                                                           InvocationTargetException {
        try {
            Method method = object.getClass().getMethod("initWithDefaults",(java.lang.Class[])null);        
            if(method!=null)
                method.invoke(object,(java.lang.Object[])null);
        }
        catch(NoSuchMethodException ex){};
    }
    
    protected Object createInstanceForElement(Class objectClass, ElementInfo elementInfo) throws InstantiationException, 
                                                                              IllegalAccessException, 
                                                                              NoSuchMethodException, 
                                                                              InvocationTargetException {
        if(objectClass.isMemberClass() && elementInfo.getParentObject()!=null) {
            Constructor decl = objectClass.getDeclaredConstructor(elementInfo.getParentObject().getClass());
            return decl.newInstance(elementInfo.getParentObject());
        }
        else {
            return objectClass.newInstance();
        }
    }
        
    protected DecodedObject decodeSequence(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception {        
        Object sequence = createInstanceForElement(objectClass,elementInfo);
        initDefaultValues(sequence);
        
        DecodedObject fieldTag = decodeTag(stream);
        int sizeOfSequence = 0;
        if(fieldTag!=null)
            sizeOfSequence+=fieldTag.getSize();
        
        Field[] fields =objectClass.getDeclaredFields();
        for(int i=0; i<fields.length; i++) {        
            Field field = fields[i];            
            DecodedObject obj = decodeSequenceField(fieldTag,sequence,field,stream,elementInfo, true);
            if(obj!=null) {
                sizeOfSequence+=obj.getSize();
                
                if(i!=fields.length-1 && fields[i+1].isAnnotationPresent(ASN1Any.class)) {
                }
                else {
                    if(i<fields.length-1) {
                        fieldTag = decodeTag(stream);
                        if(fieldTag!=null) {                        
                            sizeOfSequence+=fieldTag.getSize();
                        }
                        else
                            break;
                    }
                }
            };
        }
        return new DecodedObject(sequence,sizeOfSequence);
    }
    

    protected DecodedObject decodeSequenceField(DecodedObject fieldTag, Object sequenceObj, Field field, InputStream stream, ElementInfo elementInfo, boolean optionalCheck) throws  Exception {
        ElementInfo info = new ElementInfo(field, field.getAnnotation(ASN1Element.class));
        if(field.getType().isMemberClass()) {
            info.setParentObject(sequenceObj);
        }
        info.setGenericInfo(field.getGenericType());
        if(field.isAnnotationPresent(ASN1Null.class)) {            
            return decodeNull(fieldTag,field.getType(),info, stream);
        }
        else {
            DecodedObject value = decodeClassType(fieldTag,field.getType(),info,stream);
            if(value!=null) {                
                invokeSetterMethodForField(field, sequenceObj, value.getValue());
            }
            else {
                if(optionalCheck)
                    checkForOptionalField(field);
            }
            return value;
        }
    }

    protected DecodedObject decodeChoice(DecodedObject decodedTag,Class objectClass, ElementInfo elementInfo, InputStream stream)  throws Exception {
        Object choice = createInstanceForElement(objectClass,elementInfo);
        DecodedObject value = null;
        for ( Field field : objectClass.getDeclaredFields() ) {
            if(!field.isSynthetic()) {
                ElementInfo info = new ElementInfo(field, field.getAnnotation(ASN1Element.class));
                if(field.getType().isMemberClass()) {
                    info.setParentObject(choice);
                }                
                info.setGenericInfo(field.getGenericType());            
                value = decodeClassType(decodedTag, field.getType(),info,stream);
                if(value!=null) {
                    invokeSelectMethodForField(field, choice, value.getValue());
                    break;
                };
            }            
        }
        if(value == null && elementInfo.getElement()!=null && !elementInfo.getElement().isOptional()) {
            throw new  IllegalArgumentException ("The choice '" + objectClass.toString() + "' does not have a selected item!");
        }
        else
            return new DecodedObject(choice, value!=null ? value.getSize(): 0);
    }
        
    protected DecodedObject decodeEnum(DecodedObject decodedTag,Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception  {
        Object result = objectClass.newInstance();
        Field field = objectClass.getDeclaredField("value");

        Class enumClass = null;
        for(Class cls : objectClass.getDeclaredClasses()) {
            if(cls.isEnum()) {
                enumClass = cls;
                break;
            }
        };

        DecodedObject itemValue = decodeEnumItem(decodedTag, field.getType(),enumClass, elementInfo, stream );
        
        Field param = null;
        if(itemValue!=null) {
            for(Field enumItem: enumClass.getDeclaredFields()) {
                if(enumItem.isAnnotationPresent(ASN1EnumItem.class)) {
                    ASN1EnumItem meta = enumItem.getAnnotation(ASN1EnumItem.class);
                    if(meta.tag() == (Integer)itemValue.getValue()) {
                        param = enumItem;
                        break;
                    }
                }
            }
            invokeSetterMethodForField ( field, result, param.get(null)) ;
        }        
        return new DecodedObject(result,itemValue.getSize());
    }
        
    protected abstract DecodedObject decodeEnumItem(DecodedObject decodedTag,Class objectClass, Class enumClass, ElementInfo elementInfo, InputStream stream) throws Exception ;
    

    protected DecodedObject decodeElement(DecodedObject decodedTag,Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception  {
        elementInfo.setAnnotatedClass(objectClass);
        return decodeClassType(decodedTag, objectClass,elementInfo, stream);
    }
    
    protected DecodedObject decodeBoxedType(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception  {    
        Object resultObj = createInstanceForElement(objectClass,elementInfo);
        
        DecodedObject result = new DecodedObject(resultObj);
        
        Field field = objectClass.getDeclaredField("value");
        elementInfo.setAnnotatedClass(field);
        elementInfo.setGenericInfo(field.getGenericType());
        if(field.getType().isMemberClass()) {
            elementInfo.setParentObject(resultObj);
        }                
        
        
        DecodedObject value = null;
        if(field.isAnnotationPresent(ASN1Null.class)) {
            value = decodeNull(decodedTag, field.getType(),elementInfo,stream);
        }
        else {            
            value = decodeClassType(decodedTag, field.getType(), elementInfo, stream);
            if(value!=null) {
                result.setSize(value.getSize());
                invokeSetterMethodForField(field,resultObj,value.getValue());
            }
            else
                result = null;
        }
        return result;
    }

    protected abstract DecodedObject  decodeBoolean(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception;

    protected abstract DecodedObject  decodeAny(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception ;

    protected abstract DecodedObject  decodeNull(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception ;

    protected abstract DecodedObject  decodeInteger(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception ;

    protected abstract DecodedObject  decodeReal(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception ;

    protected abstract DecodedObject  decodeOctetString(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception ;

    protected abstract DecodedObject  decodeBitString(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception ;

    protected abstract DecodedObject  decodeString(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception ;

    protected abstract DecodedObject  decodeSequenceOf(DecodedObject decodedTag, Class objectClass, ElementInfo elementInfo, InputStream stream) throws Exception ;
    
    protected abstract DecodedObject decodeTag(InputStream stream) throws Exception ;
    
    //
}
